ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -0,0 +1,145 @@
+  collectors_cpu_and_wall_time_class = rb_define_class_under(collectors_module, "CpuAndWallTime", rb_cObject);
+
+  // Instances of the CpuAndWallTime class are "TypedData" objects.
+  // "TypedData" objects are special objects in the Ruby VM that can wrap C structs.
+  // In this case, it wraps the cpu_and_wall_time_collector_state.
+  //
+  // Because Ruby doesn't know how to initialize native-level structs, we MUST override the allocation function for objects
+  // of this class so that we can manage this part. Not overriding or disabling the allocation function is a common
+  // gotcha for "TypedData" objects that can very easily lead to VM crashes, see for instance
+  // https://bugs.ruby-lang.org/issues/18007 for a discussion around this.
+  rb_define_alloc_func(collectors_cpu_and_wall_time_class, _native_new);
+
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_initialize", _native_initialize, 3);
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_sample", _native_sample, 1);
+  rb_define_singleton_method(collectors_cpu_and_wall_time_class, "_native_thread_list", _native_thread_list, 0);
+}
+
+// This structure is used to define a Ruby object that stores a pointer to a struct cpu_and_wall_time_collector_state
+// See also https://github.com/ruby/ruby/blob/master/doc/extension.rdoc for how this works
+static const rb_data_type_t cpu_and_wall_time_collector_typed_data = {
+  .wrap_struct_name = "Datadog::Profiling::Collectors::CpuAndWallTime",
+  .function = {
+    .dmark = cpu_and_wall_time_collector_typed_data_mark,
+    .dfree = cpu_and_wall_time_collector_typed_data_free,
+    .dsize = NULL, // We don't track profile memory usage (although it'd be cool if we did!)
+    //.dcompact = NULL, // FIXME: Add support for compaction
+  },
+  .flags = RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static void cpu_and_wall_time_collector_typed_data_mark(void *state_ptr) {
+  struct cpu_and_wall_time_collector_state *state = (struct cpu_and_wall_time_collector_state *) state_ptr;
+
+  // Update this when modifying state struct
+  rb_gc_mark(state->recorder_instance);
+}
+
+static void cpu_and_wall_time_collector_typed_data_free(void *state_ptr) {
+  struct cpu_and_wall_time_collector_state *state = (struct cpu_and_wall_time_collector_state *) state_ptr;
+
+  // Update this when modifying state struct
+
+  // Important: Remember that we're only guaranteed to see here what's been set in _native_new, aka
+  // pointers that have been set NULL there may still be NULL here.
+  if (state->sampling_buffer != NULL) sampling_buffer_free(state->sampling_buffer);
+
+  xfree(state);
+}
+
+static VALUE _native_new(VALUE klass) {
+  struct cpu_and_wall_time_collector_state *state = xcalloc(1, sizeof(struct cpu_and_wall_time_collector_state));
+
+  if (state == NULL) {
+    rb_raise(rb_eNoMemError, "Failed to allocate memory for components of Datadog::Profiling::Collectors::CpuAndWallTime");
```

Comment:
Is it common practice to invoke `rb_raise` when we fail to allocate memory?
It seems to me that a single `sizeof(struct cpu_and_wall_time_collector_state)` would require less space then whatever happens inside `rb_raise(rb_eNoMemError, "Failed to allocate memory for components of Datadog::Profiling::Collectors::CpuAndWallTime");`, aka, will this `rb_raise` ever execute?

Reply:
That's a very good question. I've picked up this pattern from the ffi gem (https://github.com/ffi/ffi/search?q=rb_eNoMemError) but that doesn't mean that it's correct, just means that it's what the ffi gem does.

I went back and looked at the MRI sources and there's a `rb_memerror()`:
that is exported for gems to use and that seems to handle your concerns exactly -- it tries to raise that exception, but takes care with the fact that if we're running out of memory, things may be extremely on fire so this is not a normal exception:

https://github.com/ruby/ruby/blob/60d45b2ee86a80e248c3bff0c90c981ed2168ac3/include/ruby/internal/intern/gc.h#L42-L53

Great point. I'll update this PR and any such usages to use `rb_memerror()`.

Reply:
From https://github.com/ruby/ruby/blob/60d45b2ee86a80e248c3bff0c90c981ed2168ac3/include/ruby/internal/intern/gc.h#L42-L53:
```
tries hard to show something, but ultimately can die silently
```
2irl4meirl